docs/fsharp/language-reference/functions/recursive-functions-the-rec-keyword.md

```diff
@@ -26,15 +26,42 @@ function2-body
 
 Recursive functions, functions that call themselves, are identified explicitly in the F# language. This makes the identifer that is being defined available in the scope of the function.
 
-The following code illustrates a recursive function that computes the *n*<sup>th</sup> Fibonacci number.
+The following code illustrates a trivial recursive function that computes the *n*<sup>th</sup> Fibonacci number.
 
 [!code-fsharp[Main](~/samples/snippets/fsharp/lang-ref-1/snippet4001.fs)]
 
 > [!NOTE]
-> In practice, code like that above is wasteful of memory and processor time because it involves the recomputation of previously computed values.
+> In practice, code like the previous sample is not ideal because it unecessarily recomputes values that have already been computed. This is because it is not tail recursive, which is explained further in the document.
 
 Methods are implicitly recursive within the type; there is no need to add the `rec` keyword. Let bindings within classes are not implicitly recursive.
 
+## Tail recursion
+
+For some recursive functions, it is necessary to refactor a more "pure" definition to one that is [tail recursive](https://cs.stackexchange.com/questions/6230/what-is-tail-recursion). This prevents unecessary recomputations. For example, the previous fibonacci number generator can be rewritten like this:
+
+```fsharp
+let fib n =
+    let rec loop acc1 acc2 n =
+        match n with
+        | 0 -> acc1
+        | 1 -> acc2
+        | _ ->
+            loop acc2 (acc1 + acc2) (n - 1)
+    loop 0 1 n
+```
+
+This is a more complicated implementation. Generating a fibonacci number is a great example of a "naiive" algorithm that is mathematically pure, but inefficient in practice. Making it efficient in F# while still remaining recursively-defined makes use of a few things:
+
+* A recursive inner function named `loop`, which is an idiomatic F# pattern
+* Two accumulator parameters, which pass accumulate values to recursive calls
+* A check on the value of `n` to return a specific accumulate
```

Comment:
Not sure if these should end in a period or not.

Reply:
I generally tend not to place periods here, but if that's the style guide it can be done.

Reply:
I'm not sure what's the style-guide in this case.

@gewarren Do you know?